src/core/ddsi/src/ddsi_typelib.c

```diff
@@ -1038,6 +1038,10 @@ static dds_return_t add_type_info_hash_deps_r (struct ddsi_domaingv *gv, const d
 
       case DDSI_TYPEID_KIND_FULLY_DESCRIPTIVE:
         break;
+
+      case DDSI_TYPEID_KIND_INVALID:
+        ret = DDS_RETCODE_BAD_PARAMETER;
```

Comment:
Good point, the code is indeed somewhat fishy.

If I'm not mistaken `ddsi_typeid_fini(x) ; ddsi_typeid_fini(x)` is allowed because you end up in `dds_stream_free_sample`, which nulls the pointers it freed and therefore calling it a second time will be a no-op. That suggests the `ddsi_typeid_fini (&tmpl.src_type_id)` is harmless for the minimal/complete cases.

Secondly, on a cursory glance, all callers of `add_type_info_hash_deps_r` free the `type_info` on error, except for recursive calls where it simply passes it unchanged, so that seems okay, too.

In short I wouldn't be surprised if it ended up doing the right thing, but if so, it does it in a manner so roundabout that I don't feel comfortable with it. In short, I'll go double-check the code paths and see if I can eliminate the weirdness without harming (apparent) readability 🙂